components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1306 @@
+use std::convert::{TryFrom, TryInto};
+
+use crate::codec;
+use crate::codec::mysql;
+use crate::codec::TEN_POW;
+use crate::codec::{Error, Result};
+use crate::expr::{EvalContext, Flag, SqlMode};
+
+use tidb_query_datatype::FieldTypeTp;
+
+use bitfield::bitfield;
+use chrono::prelude::*;
+
+const MIN_TIMESTAMP: i64 = 0;
+const MAX_TIMESTAMP: i64 = (1 << 31) - 1;
+const MICRO_WIDTH: usize = 6;
+
+fn is_leap_year(year: u32) -> bool {
+    year & 3 == 0 && (year % 100 != 0 || year % 400 == 0)
+}
+
+fn last_day_of_month(year: u32, month: u32) -> u32 {
+    match month {
+        4 | 6 | 9 | 11 => 30,
+        2 => is_leap_year(year) as u32 + 28,
+        _ => 31,
+    }
+}
+
+fn assert(expected: bool) -> Result<()> {
```

Comment:
Too ugly.. In addition if you call it assert then it should panic.

Reply:
Introduced `boolinator` and removed it.